x-pack/plugin/esql/compute/src/test/java/org/elasticsearch/compute/operator/OperatorTestCase.java

```diff
@@ -73,69 +73,58 @@ public final void testSimpleLargeInput() {
     }
 
     /**
-     * A {@link ByteSizeValue} that is small enough that running {@link #simple}
-     * on {@link #simpleInput} will exhaust the breaker and throw a
-     * {@link CircuitBreakingException}. We should make an effort to make this
-     * number as large as possible and still cause a break consistently so we get
-     * good test coverage. If the operator can't break then throw an
-     * {@link AssumptionViolatedException}.
+     * Enough memory for {@link #simple} not to throw a {@link CircuitBreakingException}.
+     * It's fine if this is <strong>much</strong> more memory than {@linkplain #simple} needs.
+     * When we want to make {@linkplain #simple} throw we'll find the precise amount of memory
+     * that'll make it throw with a binary search.
      */
-    protected abstract ByteSizeValue memoryLimitForSimple();
-
-    /**
-     * Run {@link #simple} with a circuit breaker limited to somewhere
-     * between 0 bytes and {@link #memoryLimitForSimple} and assert that
-     * it breaks in a sane way.
-     */
-    public final void testSimpleCircuitBreaking() {
-        testSimpleCircuitBreaking(ByteSizeValue.ofBytes(randomLongBetween(0, memoryLimitForSimple().getBytes())));
+    protected ByteSizeValue enoughMemoryForSimple() {
+        return ByteSizeValue.ofGb(1);
     }
 
     /**
-     * Run {@link #simple} with a circuit breaker configured limited to
-     * {@link #memoryLimitForSimple} and assert that it breaks in a sane way.
-     * <p>
-     *     This test helps to make sure that the limits set by
-     *     {@link #memoryLimitForSimple} aren't too large.
-     *     {@link #testSimpleCircuitBreaking}, with it's random configured
-     *     limit will use the actual maximum very rarely.
-     * </p>
+     * Run {@link #simple} with a circuit breaker NOCOMMIT
      */
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/103789")
-    public final void testSimpleCircuitBreakingAtLimit() {
-        testSimpleCircuitBreaking(memoryLimitForSimple());
+    public final void testSimpleCircuitBreaking() {
+        ByteSizeValue memoryLimitForSimple = enoughMemoryForSimple();
+        Operator.OperatorFactory simple = simple();
+        DriverContext inputFactoryContext = driverContext();
+        List<Page> input = CannedSourceOperator.collectPages(simpleInput(inputFactoryContext.blockFactory(), between(1_000, 10_000)));
+        try {
+            ByteSizeValue limit = BreakerTestUtil.findBreakerLimit(
+                memoryLimitForSimple,
+                l -> runWithLimit(simple, CannedSourceOperator.deepCopyOf(input), l)
```

Comment:
This means we need to execute runWithLimit approximately 15 times to determine the lower bound if it's less than 100KB. Should we consider reducing the upper bound? However, I am fine with leaving it as is if there's no significant increase in time for CI.

Reply:
It doesn't seem to mind too much. On my laptop running it for all 49 existing tests takes 8.6 seconds. That seems ok.

Reply:
Here's the logging of the slowest one:
```
[2024-01-08T11:43:05,418][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 1gb. 1gb bytes remaining.
[2024-01-08T11:43:05,473][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 512mb. 512mb bytes remaining.
[2024-01-08T11:43:05,528][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 256mb. 256mb bytes remaining.
[2024-01-08T11:43:05,581][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 128mb. 128mb bytes remaining.
[2024-01-08T11:43:05,637][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 64mb. 64mb bytes remaining.
[2024-01-08T11:43:05,692][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 0b and 32mb. 32mb bytes remaining.
[2024-01-08T11:43:05,696][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 16mb and 32mb. 16mb bytes remaining.
[2024-01-08T11:43:05,701][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 32mb. 8mb bytes remaining.
[2024-01-08T11:43:05,759][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 28mb. 4mb bytes remaining.
[2024-01-08T11:43:05,815][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 24mb and 26mb. 2mb bytes remaining.
[2024-01-08T11:43:05,820][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25mb and 26mb. 1mb bytes remaining.
[2024-01-08T11:43:05,825][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 26mb. 512kb bytes remaining.
[2024-01-08T11:43:05,880][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.7mb. 256kb bytes remaining.
[2024-01-08T11:43:05,937][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.6mb. 128kb bytes remaining.
[2024-01-08T11:43:05,996][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 64kb bytes remaining.
[2024-01-08T11:43:06,049][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 32kb bytes remaining.
[2024-01-08T11:43:06,104][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 16kb bytes remaining.
[2024-01-08T11:43:06,109][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 8kb bytes remaining.
[2024-01-08T11:43:06,114][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 4kb bytes remaining.
[2024-01-08T11:43:06,172][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 2kb bytes remaining.
[2024-01-08T11:43:06,225][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 1kb bytes remaining.
[2024-01-08T11:43:06,231][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 512b bytes remaining.
[2024-01-08T11:43:06,284][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 256b bytes remaining.
[2024-01-08T11:43:06,338][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 128b bytes remaining.
[2024-01-08T11:43:06,393][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 64b bytes remaining.
[2024-01-08T11:43:06,398][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 32b bytes remaining.
[2024-01-08T11:43:06,403][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 16b bytes remaining.
[2024-01-08T11:43:06,408][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 8b bytes remaining.
[2024-01-08T11:43:06,414][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 4b bytes remaining.
[2024-01-08T11:43:06,420][INFO ][o.e.t.BreakerTestUtil    ] [testSimpleCircuitBreaking] Between 25.5mb and 25.5mb. 2b bytes remaining.
[2024-01-08T11:43:06,485][INFO ][o.e.c.a.CountDistinctLongGroupingAggregatorFunctionTests] [testSimpleCircuitBreaking] testing with 3.3mb against a limit of 25.5mb
[2024-01-08T11:43:06,487][INFO ][o.e.c.a.CountDistinctLongGroupingAggregatorFunctionTests] [testSimpleCircuitBreaking] after test
```

Seems ok to me.

Reply:
Yes, let's leave it as is.